### ISSUE TITLE
Better IS_BROWSER check

### ISSUE THREAD

**Issue by danieldunderfelt** [styled-components/styled-components#1598]
I am working on a server-side rendered project (with Razzle) that uses [node-pixi](https://github.com/timknip/node-pixi) for rendering Pixi scenes to an image on the server. When I tried to start using styled.components, I got the error `TypeError: document.querySelectorAll is not a function` as styled-components tried to rehydrate the styles... on the server.

I investigated, and it turns out that styled-components uses a very simple `typeof window !== 'undefined'` to determine whether or not the current environment is a browser. It also turns out that, in order to make PIXI work server-side, node-pixi assigns a shimmed Window object onto `global.window`, making styled-components think that the server is a browser.

You may argue that node-pixi should change. There are a few reasons why it's better to change styled-components:

- It's much easier, since we can just change one line of code. Who knows to what extend PIXI uses the window object.
- styled-components is arguably the more "general" library, and I assume that the maintainers want it to work in a variety of weird situations.

So I propose the following change, in `src/constants.js` on [line 9](https://github.com/styled-components/styled-components/blob/19c98575466218b8f7a0110bc13ed2844519da91/src/constants.js#L9):

```
// Instead of:
export const IS_BROWSER = typeof window !== 'undefined'

// Do this:
export const IS_BROWSER = typeof window !== 'undefined' && window.window === window
```

To shim the window, which may be required in certain cases like PIXI's, you hardly need to do the circular assignment. That's why this is a much safer test that will make styled-components work more reliably.

What do you think? I might make a PR since I'm gonna fork SC now and get my app working again.

EDIT: As I was implementing this, the original `this.window === this` didn't pass the tests, so I settled on a similar solution that does pass the tests and works as intended.

**Comment by mxstbr** [comment]
I like that!

**Comment by danieldunderfelt** [comment]
@mxstbr Cool! I'm trying to implement a better check now, but it's not as easy as I thought it'd be when I wrote the issue. I'll update the issue and make a PR when I have something that works!

**Comment by kitten** [comment]
Resolved by #1599 and released in https://github.com/styled-components/styled-components/releases/tag/v3.2.2
